librad/src/id.rs

```diff
@@ -0,0 +1,273 @@
+use hex::{decode, encode};
+use multihash::{Multihash, Sha2_256};
+use std::collections::{HashMap, HashSet};
+//use serde::{Deserialize, Serialize};
+//use serde_json;
+use crate::{
+    keys::device::{Key, PublicKey, Signature},
+    peer::PeerId,
+};
+
+pub mod data;
+
+#[derive(Debug, Fail)]
+pub enum Error {
+    #[fail(display = "Serialization failed")]
+    SerializationFailed(serde_json::error::Error),
+
+    #[fail(display = "Invalid UTF8")]
+    InvalidUtf8(std::string::FromUtf8Error),
+
+    #[fail(display = "Invalid hex")]
+    InvalidHex(String),
+
+    #[fail(display = "Invalid hash")]
+    InvalidHash(String),
+
+    #[fail(display = "Signature already present")]
+    SignatureAlreadyPresent(PublicKey),
+
+    #[fail(display = "Key not present")]
+    KeyNotPresent(PublicKey),
+
+    #[fail(display = "User not present")]
+    UserNotPresent(RadicleUri),
+
+    #[fail(display = "User key not present")]
+    UserKeyNotPresent(RadicleUri, PublicKey),
+
+    #[fail(display = "Signature missing")]
+    SignatureMissing,
+
+    #[fail(display = "Signature decoding failed")]
+    SignatureDecodingFailed,
+
+    #[fail(display = "Signature verification failed")]
+    SignatureVerificationFailed,
+
+    #[fail(display = "Resolution failed")]
+    ResolutionFailed(String),
+}
+
+#[derive(Clone, Debug)]
+pub struct RadicleUri {
+    hash: Multihash,
+    /* pub repo: Multihash,
+     * pub root: Multihash,
+     * pub branch: String,
+     * pub file: String, */
+}
+
+impl RadicleUri {
+    pub fn new(hash: Multihash) -> Self {
+        Self { hash }
+    }
+    pub fn hash(&self) -> &Multihash {
+        &self.hash
+    }
+
+    pub fn from_str(s: &str) -> Result<Self, Error> {
+        let bytes = decode(s).map_err(|_| Error::InvalidHex(s.to_owned()))?;
+        let hash = Multihash::from_bytes(bytes).map_err(|_| Error::InvalidHash(s.to_owned()))?;
+        Ok(Self { hash })
+    }
+}
+
+impl ToString for RadicleUri {
+    fn to_string(&self) -> String {
+        encode(self.hash.to_vec())
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum Signatory {
+    User(RadicleUri),
+    OwnedKey,
+}
+
+#[derive(Clone, Debug)]
+pub struct EntitySignature {
+    pub by: Signatory,
+    pub sig: Signature,
+}
+
+pub trait Resolver<E>
+where
+    E: Entity,
+{
+    fn resolve(&self, uri: &RadicleUri) -> Result<E, Error>;
+}
+
+pub trait RevisionsResolver<E>
+where
+    E: Entity,
+{
+    fn resolve_revisions(&self, uri: &RadicleUri) -> Box<dyn Iterator<Item = &E>>;
+}
+
+pub trait Entity {
```

Comment:
As an organisational thing, I think it would be better to have these traits in one module and the data types in another :)

What do you think?